src/Str.php

```diff
@@ -0,0 +1,77 @@
+<?php
+
+namespace OL\String;
+
+use OL\String\Contracts\Str as StrContract;
+
+/**
+ * Class Str.
+ *
+ * Basic manipulation of strings.
+ */
+class Str implements StrContract
+{
+    /**
+     * @var string Raw string content
+     */
+    protected $payload;
+
+    /**
+     * Str constructor.
+     *
+     * The constructor will enforce string type, since it can also receive
+     * another Str object (self) as it can be casted to string.
+     *
+     * @param string|StrContract $payload The actual string
+     */
+    public function __construct(string $payload)
+    {
+        // casts the payload as string and them assign on class scope
+        $this->payload = (string) $payload;
+    }
+
+    /**
+     * Str objects should should be able to be cast as string.
+     *
+     * @return string The actual string
+     */
+    public function __toString()
+    {
+        return $this->contents();
+    }
+
+    /**
+     * Public method to retrieve the raw string contents of an object.
+     *
+     * @return string
+     */
+    public function contents()
+    {
+        return $this->payload;
+    }
+
+    /**
+     * Replace one or more entries from the given object with a given value
+     *
+     * @param array|string $search What to search for
+     * @param string $replace Replace with?
+     *
+     * @return StrContract Object instance with the replaced text.
+     */
+    public function replace($search, string $replace)
```

Comment:
I am not so sure with "PHP strings are mutable" because they are primitives. Primitives have no state, they're just a value. When you change a string variable, you don't mutate the value, you just replace it with a new one. But when I said that immutability is the same implementation of native PHP functions, I refer to the fact that common string functions like `str_replace` return a new value.

I understand the point about performance. And also agree if we're going to have both "Str" and "StrImmutable" (I don't like the `ImmutableStr` name).

Another point that I want to show is template strings:

```php
<?php

$tpl = new Str('[Person] love PHP-OL!');

foreach (['naroga', 'marcossffilho', 'hernandev', 'everyone'] as $person) {
  echo $tpl->replace('[Person]', $person);
}
```